### PR TITLE
Fixed 6287 - Added support for pause/no-pause in pipeline config API

### PR DESCRIPTION
**Pull request by msaurabh1978**
Fixed Issue: #6287 (please refer to this issue for any design related details)

Implementation details are,

- [x] Updated create pipeline config API to honor request header `X-pause-pipeline-after-create` to pause or un-pause the pipeline after creation

- [x] Added unit tests for pipeline config controller v7

### CODE REVIEW

File: api/api-pipeline-config-v7/src/main/java/com/thoughtworks/go/apiv7/admin/pipelineconfig/PipelineConfigControllerV7.java [hunk 1]
```diff
@@ -131,6 +139,10 @@ public String create(Request req, Response res) {
         Username userName = SessionUtils.currentUsername();
         pipelineConfigService.createPipelineConfig(userName, pipelineConfigFromRequest, result, group);
 
+        if(shouldPausePipelineOnCreate(req)){
+            pipelinePauseService.pause(pipelineConfigFromRequest.name().toString(), null, userName);
```
Comment (maheshp) [inline api/api-pipeline-config-v7/src/main/java/com/thoughtworks/go/apiv7/admin/pipelineconfig/PipelineConfigControllerV7.java]: Should we add a pause message for this?

Comment (ketan) [reply api/api-pipeline-config-v7/src/main/java/com/thoughtworks/go/apiv7/admin/pipelineconfig/PipelineConfigControllerV7.java]: And default to a sensible message, if one is not specified.

**Comment by ketan** [comment]
Question: should this be part of a new API version? Specifying `X-pause-pipeline-after-create` on api v7 but with an older version of the gocd server will have no effect.

File: api/api-pipeline-config-v7/src/main/java/com/thoughtworks/go/apiv7/admin/pipelineconfig/PipelineConfigControllerV7.java (hunk 1 above)
Comment (msaurabh1978) [reply api/api-pipeline-config-v7/src/main/java/com/thoughtworks/go/apiv7/admin/pipelineconfig/PipelineConfigControllerV7.java]: @ketan - Are u talking about making changes in <code>PipelinePauseService.pause()</code> method? wherein if caller of that method, does not specify the pause cause, then the sensible default message is used instead. That way pause message is never null in the database after this change. Is that right?

**Comment by msaurabh1978** [comment]
> Question: should this be part of a new API version? Specifying `X-pause-pipeline-after-create` on api v7 but with an older version of the gocd server will have no effect.

Thinking aloud - in what case(s) then would we ever make changes to existing API version? because based on what you said (older version of GoCD server will have no effect) would be valid for any kind of change?

**Comment by maheshp** [comment]
> Thinking aloud - in what case(s) then would we ever make changes to existing API version? because based on what you said (older version of GoCD server will have no effect) would be valid for any kind of change?

I don't remember we changing an existing API. A change to the API(contract) has always lead to a new version of the API.

**Comment by arvindsv** [comment]
I'd expect us to change existing APIs only when fixing bugs in the API. If there's a new feature added or old feature removed, we should bump it up.

**Comment by msaurabh1978** [comment]
Can u pls merge this pull request? I have incorporated all your review comments.

**Comment by msaurabh1978** [comment]
Below is how the pause message shows up for a paused pipeline through pipeline config API.

**Comment by GaneshSPatil** [comment]
@msaurabh1978 -- When we create a pipeline from UI, it shows pause message as `Under Construction`.

Can we use the same message:
1. As it is short and descriptive.
2. Consistency with the message that gets shown when we create a pipeline from UI.

**Comment by msaurabh1978** [comment]
> @msaurabh1978 -- When we create a pipeline from UI, it shows pause message as `Under Construction`.
> 
> Can we use the same message:
> 
> 1. As it is short and descriptive.
> 2. Consistency with the message that gets shown when we create a pipeline from UI.

Yes, we can.

The way I am seeing it is `pause-cause` should reflect the actual reason for pausing the pipeline at any point in time (whether paused on creating or paused at a later time). `Under Construction` totally makes sense in case of pipeline getting created from UI because, at that point, it is really under construction. In this particular case, it is paused to have more control over when it can be started after reviewing it.

Another point is, whenever pipeline is paused at a later time after its creation, then the user can specify any message while pausing (could be a longer message as well).

So I think we need not be consistent with pause message on creating (as the context for pausing is different). Does this make sense?

**Comment by maheshp** [comment]
There could be various reasons for pausing a pipeline post creation( to start later, to review and edit further, etc...) I think we should have a generic message that its paused post creation, how bout saying `This pipeline has been paused after creating it through API.`

**Comment by ketan** [comment]
> how bout saying `This pipeline has been paused after creating it through API.`

Can I suggest passing in another header with a pause cause, and if that header is blank, then fallback to a sensible default?

**Comment by ketan** [comment]
Discussed with @msaurabh1978 about the following options for default pause message, if one is not specified:

- @GaneshSPatil's suggestion "Under construction" does not clearly spell out why
- @maheshp's suggestion "This pipeline has been paused after creating it through API" makes it look like that's the default behavior, and gocd will pause all pipelines going forward
- @msaurabh1978's suggestion "intentionally paused" also makes it look like it's a decision that gocd made on behalf of the user.

We decided to instead use `No pause cause was specified when pipeline was created via API`. This seems to clarify exactly what has happened, without making it look like the GoCD server is making any decisions.

**Comment by MRajasekhar** [comment]
Hi All, 

we are in situation to create pipelines with pause state from config repository code. the way we create pipelines from yaml file and config repo plugin reads -> creates/modify/deletes pipeline. 
When we are creating new pipelines - it is becoming active and running. we need to have it pause state so that we can release when it is ready to avoid unnecessary failures. 

Regards,
Rajasekhar

**Pull request merged**